### PR TITLE
Refactor Env and Interfaces from ClientConfig base class

### PR DESCRIPTION
**Pull request by rihp**
Initially we had a ClientConfig base class with a list of Environments, and a list of interfaces,

```py
@dataclass(slots=True, kw_only=True)
class ClientConfig:
    envs: List[Env] = field(default_factory=list)
    interfaces: List[InterfaceImplementations] = field(default_factory=list)
    resolver: IUriResolver
```

Both of these fields are now being changed to a Dictionary for simpler client configuration.

```py
@dataclass(slots=True, kw_only=True)
class ClientConfig:
    envs: Dict[Uri, Env: Dict[str, Any]] = field(default_factory=dict) 
    interfaces: Dict[Uri, List[Uri]]
    resolver: IUriResolver
```

This changes break some of the client functionality, working on those fixes now

-------

### Status update 10/nov/2022:

Was able to get all tests from Client package passing but broke some other packages like msgpack and core at build time.
I'm fixing those up next

### Status update 11/nov/2022:
- [x] env refactor is building properly after all changes! 
- [x] interfaces and interfacer implementations have been refactor working and merged into this branch!

### CODE REVIEW

**Comment by rihp** [comment]
The build is now passing, all green after the Env interface refactor 😎 

now will look at the interfaces refactor

**Comment by rihp** [comment]
Included now the `Interface Implementations` refactor on this branch..!
there is still this Type error showing up on vscode, yet it's building properly! 
image

File: packages/polywrap-msgpack/polywrap_msgpack/__init__.py [hunk 1]
```diff
@@ -52,10 +52,12 @@ def sanitize(value: Any) -> Any:
     """
     if isinstance(value, dict):
         dictionary: Dict[Any, Any] = value
-        for key, val in dictionary.items():
+        for key, val in list(dictionary.items()):
             if isinstance(key, str):
                 dictionary[key] = sanitize(val)
-            else:
+            elif key.uri:
```
Comment (dOrgJelli) [inline packages/polywrap-msgpack/polywrap_msgpack/__init__.py:58]: I'm not familiar with this codebase, but why do we want to check whether the `key.uri` exists here? This msgpack package should be independent of any of the polywrap specific semantics, so referencing "uri" doesn't quite make sense here to me.

File: packages/polywrap-core/polywrap_core/utils/get_env_from_uri_history.py [hunk 2]
```diff
@@ -1,10 +1,10 @@
-from typing import List, Union
+from typing import List, Union, Dict, Any
 
 from ..types import Client, Env, Uri
 
 
 def get_env_from_uri_history(
     uri_history: List[Uri], client: Client
-) -> Union[Env, None]:
+) -> Union[Dict[str, Any], None]:
```
Comment (dOrgJelli) [inline packages/polywrap-core/polywrap_core/utils/get_env_from_uri_history.py:8]: Why not use the `Env` type here? It helps me better understand what this code is doing.

**Review by dOrgJelli — CHANGES_REQUESTED** [review]

File: packages/polywrap-client/polywrap_client/client.py [hunk 3]
```diff
@@ -51,24 +54,35 @@ def get_uri_resolver(
...
+        interfaces: Dict[Uri, List[Uri]] = self._config.interfaces
+        return interfaces
+
+    def get_implementations(self, uri: Uri) -> Result[Union[List[Uri], None] ]:
+        interfaces: Dict[Uri, List[Uri]] = self.get_interfaces()
+        if interfaces.get(uri):
+            print(f"{type(interfaces)=}")
+            print(f"{interfaces=}")
+            print(f"{interfaces.get(uri)=}")
+
```
Comment (Niraj-Kamdar) [inline packages/polywrap-client/polywrap_client/client.py:71]: Remove debug leftover

File: packages/polywrap-core/polywrap_core/types/client.py [hunk 4]
```diff
@@ -45,17 +46,17 @@ class GetManifestOptions(DeserializeManifestOptions):
 
 class Client(Invoker, UriResolverHandler):
     @abstractmethod
-    def get_interfaces(self) -> List[InterfaceImplementations]:
+    def get_interfaces(self) -> Dict[Uri, List[Uri]]:
         pass
 
     @abstractmethod
-    def get_envs(self, options: Optional[GetEnvsOptions] = None) -> List[Env]:
+    def get_envs(self, options: Optional[GetEnvsOptions] = None) -> Union[Dict[Uri, Dict[str, Any]], None]:
```
Comment (Niraj-Kamdar) [inline packages/polywrap-core/polywrap_core/types/client.py:53]: change: 
`def get_envs(self, options: Optional[GetEnvsOptions] = None) -> Union[Dict[Uri, Dict[str, Any]], None]:`
to 
`def get_envs(self, options: Optional[GetEnvsOptions] = None) -> Dict[Uri, Env]:`

File: packages/polywrap-core/polywrap_core/types/client.py [hunk 5]
```diff
@@ -45,17 +46,17 @@ class GetManifestOptions(DeserializeManifestOptions):
...
         pass
 
     @abstractmethod
     def get_env_by_uri(
         self, uri: Uri, options: Optional[GetEnvsOptions] = None
-    ) -> Union[Env, None]:
```
Comment (Niraj-Kamdar) [inline packages/polywrap-core/polywrap_core/types/client.py:58]: The previous one would be better

File: packages/polywrap-core/polywrap_core/types/interface_implementation.py [hunk 6]
```diff
@@ -5,7 +5,7 @@
 
 from .uri import Uri
 
-
+# TODO:  Should we remove this interfaceimplementation?
 @dataclass(slots=True, kw_only=True)
 class InterfaceImplementations:
     interface: Uri
```
Comment (Niraj-Kamdar) [inline packages/polywrap-core/polywrap_core/types/interface_implementation.py:11]: let's remove this

File: packages/polywrap-msgpack/polywrap_msgpack/__init__.py (hunk 1 above)
Comment (Niraj-Kamdar) [reply packages/polywrap-msgpack/polywrap_msgpack/__init__.py:58]: It shouldn't be done like this, I will update it.

File: packages/polywrap-client/polywrap_client/client.py [hunk 7]
```diff
@@ -51,24 +54,35 @@ def get_uri_resolver(
...
-    ) -> Union[Env, None]:
-        return next(filter(lambda env: env.uri == uri, self.get_envs()), None)
+
+    def get_env_by_uri(self, uri: Uri, options: Optional[GetEnvsOptions] = None
+    ) -> Union[Dict[str, Any], None]:
+        envs = self.get_envs()
+        if envs is not None:
+            if hasattr(envs, 'get'):
+                result = envs.get(uri)
+                return result
```
Comment (Niraj-Kamdar) [inline packages/polywrap-client/polywrap_client/client.py:83]: use config instead

**Review by Niraj-Kamdar — CHANGES_REQUESTED** [review]
I will make the requested changes and some refactor

**Review by Niraj-Kamdar — APPROVED** [review]

**Pull request merged**